snapshots/helpers_test.go

```diff
@@ -0,0 +1,144 @@
+package snapshots_test
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	db "github.com/tendermint/tm-db"
+
+	"github.com/cosmos/cosmos-sdk/snapshots"
+)
+
+func checksum(b []byte) []byte {
+	hash := sha256.Sum256(b)
```

Comment:
These are mostly used as checksums, and cannot be trusted anyway (the snapshot metadata is off-chain and thus can be forged by anyone). Light client verification is used for the final verification, this is just a best-effort checksum to guard against IO issues and non-determinism.

Reply:
If you don't care about preimage attack resistance then don't use cryptographic hash. It's an overkill. Murmur3 hash or Googles' fast-hash are much better for some checksum. However they are easy to attack (if you want to find different data with same checksum).

Reply:
We're already using SHA-256 for everything else (including consensus-critical stuff), so I figured we might as well just use that here too. But if you'd like to use a different algo just let me know and I'll swap it.

Reply:
If we don't need preimage resistance then yes - let's use `murmur3`. Using `sha256` for hashing many big chunks of data is overkill.

Reply:
Sure. We don't need preimage resistance for safety (the light client verification handles that), but I suppose it could protect against DoS attacks - if an adversary can provide a known-bad chunk that matches the hash of a valid snapshot then they could crash the recovery process, preventing the node from state syncing.

Reply:
That should be a slashing behavior. But I guess it won't work because not every full node is bonded / active validator.

I think even with sha2 a node can send bad data and it will be only noticed after de-serialization, correct? So cryptography hash function doesn't help.

Reply:
Yes, collisions are possible with every hash function, it's just a matter of probability/difficulty.

Reply:
I didn't talk about collisions. I said that even with `sha2` a node can send a bad data:

1. Node A (bad) is creating a chunk, do some random change and then hashes it with `sha2`.
2. Node B (victim) is reconstructing a chunk, hashes it. Hash check is correct.
3. Node B tries to restore the chunk - it will eventually crash because it's broken.

Reply:
Ah, but we don't necessarily get the chunk hashes from A, they come from the snapshot manifest.

When starting state sync, Tendermint will discover snapshots in the network, by fetching manifests that include chunk hashes (in the SDK's case). It then determines which snapshot is "best" and downloads chunks from all nodes that claim to have this snapshot. If a node can generate a faulty chunk with the same hash listed in the manifest they can cause the node to fail.

The heuristic for "best" snapshot is currently very naïve (pick the newest one), but should be improved to use the one available on a majority of nodes (weighted by age). In this scenario, an adversary would have to control a majority of snapshotting nodes to be able to send a faulty chunk, but this no longer holds if the hash is vulnerable to preimage attacks.

Reply:
OK, thanks for explanation. 
So, let's keep it as is and create a new issue to see what will be the best hash function for this.

Reply:
https://github.com/cosmos/cosmos-sdk/issues/7215